flink/src/main/java/org/apache/iceberg/flink/source/StreamingReaderOperator.java

```diff
@@ -0,0 +1,220 @@
+ */
+public class StreamingReaderOperator extends AbstractStreamOperator<RowData>
+    implements OneInputStreamOperator<FlinkInputSplit, RowData> {
+
+  private static final long serialVersionUID = 1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingReaderOperator.class);
+  private final MailboxExecutorImpl executor;
+  private FlinkInputFormat format;
+
+  private transient SourceFunction.SourceContext<RowData> readerContext;
+
+  private transient ListState<FlinkInputSplit> checkpointState;
+  private transient Queue<FlinkInputSplit> splits;
+
+  private StreamingReaderOperator(
+      FlinkInputFormat format, ProcessingTimeService timeService, MailboxExecutor mailboxExecutor) {
+    this.format = Preconditions.checkNotNull(format, "The InputFormat should not be null.");
+    this.processingTimeService = timeService;
+    this.executor = (MailboxExecutorImpl) Preconditions.checkNotNull(
+        mailboxExecutor, "The mailboxExecutor should not be null.");
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    super.initializeState(context);
+
+    checkpointState = context.getOperatorStateStore().getListState(
+        new ListStateDescriptor<>("splits", new JavaSerializer<>()));
+
+    int subtaskIdx = getRuntimeContext().getIndexOfThisSubtask();
+    splits = Lists.newLinkedList();
+    if (context.isRestored()) {
+      LOG.info("Restoring state for the {} (taskIdx={}).", getClass().getSimpleName(), subtaskIdx);
+
+      for (FlinkInputSplit split : checkpointState.get()) {
+        splits.add(split);
+      }
+    }
+
+    this.readerContext = StreamSourceContexts.getSourceContext(
+        getOperatorConfig().getTimeCharacteristic(),
+        getProcessingTimeService(),
+        new Object(), // no actual locking needed
+        getContainingTask().getStreamStatusMaintainer(),
+        output,
+        getRuntimeContext().getExecutionConfig().getAutoWatermarkInterval(),
+        -1);
+
+    if (!splits.isEmpty()) {
+      enqueueProcessSplits();
+    }
+  }
+
+  @Override
+  public void processElement(StreamRecord<FlinkInputSplit> element) {
+    splits.offer(element.getValue());
+    enqueueProcessSplits();
+  }
+
+  private void enqueueProcessSplits() {
+    executor.execute(this::processSplits, this.getClass().getSimpleName());
+  }
+
+  private void processSplits() throws IOException {
+    do {
+      FlinkInputSplit split = splits.poll();
+      if (split == null) {
+        return;
+      }
+
+      LOG.debug("load split: {}", split);
+      format.open(split);
+
+      RowData nextElement = null;
+      while (!format.reachedEnd()) {
+        nextElement = format.nextRecord(nextElement);
+        if (nextElement != null) {
+          readerContext.collect(nextElement);
+        } else {
+          break;
+        }
+      }
+
+      format.close();
+    } while (executor.isIdle());
+    enqueueProcessSplits();
```

Comment:
Why does this submit itself again?

Reply:
That's a good question.   This method `processSplits` will emit all records from one `CombinedScanTask` to downstream and then yield the current thread (run out of the `while` loop ) &  re-enqueue the `processSplits` in `Mailbox` (Which is similar to the producer-consumer model, the caller do the enqueue then the single thread `executor` to consume from it).   

We will always yield from the `executor` thread after it has processed a split and then re-enqueue because we don't want to hold the __checkpointLock__ too long  (step.1) so that flink runtime won't wait long-time to trigger a new checkpoint (step.2) .   Otherwise if we hold the __checkpointLock__  and then process the splits one by one in the same `processSplits`,  finally the checkpoint will timeout so many times.  In an extreme case,  the flink streaming job will never checkpoint successfully if the splits data is huge.  

(Notice, for step.1,  when processing splits in `executor` thread we will be granted the __checkpointLock__,  that means nobody can start a new checkpoint unless the `executor` finished the current split processing;  for step.2     anybody who want to trigger a new checkpoint will need to grab the __checkpointLock__ firstly )

Reply:
I agree that the checkpoint lock should not be held for more than one input split.

This still doesn't make much sense to me. When a new split is added, `enqueueProcessSplits` runs, so the number of times `processSplits` is called should match the number of input splits. If there is a concern about executors and splits not aligning, then the mailbox executor should receive a task like `split -> processSplit(split)` so that we can ensure the mailbox has exactly one task per split.

Either way, the logic here doesn't seem to do what you're describing. Calling `enqueueProcessSplits` again probably submits a task that won't actually process any data because tasks are already created for each incoming split.

I also don't see where the checkpoint lock is passed to the executor, so how is the executor holding the checkpoint lock?

Last, why does this loop until `executor.isIdle`? Isn't this what is running in the mailbox loop? I interpret `isIdle` to be true when there is not another task for the executor to process. So if there isn't another call to `processSplits`, this one will keep running? But that contradicts the idea of one split processed at a time -- assuming I'm missing something about the lock not being passed to the mailbox executor. Alternatively, if there is another task for the executor, the loop will exit so it can run, but then enqueue another run. The end result seems like adding increasingly more tasks to the mailbox executor's queue because each task replaces itself and `processElement` keeps adding more.

I think that the problem is that the mailbox executor holds its own state queue, but we want to keep unprocessed splits in the `splits` queue so they can be included in checkpoint state. This can't add a task for each split to the mailbox because the mailbox task queue isn't part of state. I think the solution is to have a single thread that runs a loop that polls the queue and handles the checkpoint lock. That should be simpler than the mailbox executor architecture, while still handling one split at a time in a single thread and holding the checkpoint lock.

Reply:
Thanks for the careful reviewing.  After checked the code carefully again,  I think your concern is reasonable because the current implementation did not handle the issue correctly. 

Let me re-describe the problem to be solved so far.  I think everyone could design the following code in straightforward when there's a new coming split: 

```java
  @Override
  public void processElement(StreamRecord<FlinkInputSplit> element) {
    FlinkInputSplit split = element.getValue();

    LOG.debug("Start to process the split: {}", split);
    format.open(split);

    try {
      RowData nextElement = null;
      while (!format.reachedEnd()) {
        nextElement = format.nextRecord(nextElement);
        sourceContext.collect(nextElement);
      }
    } finally {
      format.close();
    }
  }
```

It's quite simple and straightforward, but we can not do that.  Because when reading the current `split1` in `processElement`,  there will be more splits coming from upstream operator,  such as `split2`, `split3`, `split4`.    Those pending splits will buffered in the flink's network queue,  it won't process the next split unless the previous one has been finished.  At this time,  if there's a checkpoint barrier coming ( Message requesting to create a new checkpoint snapshot),  then the network queue will be : `split2`, `split3`, `split4`, `checkpoint-barrier`.  The checkpoint snapshot could not start unless all the `split2`, `split3`, `split4` has been processed in this `processElement`.   Finally,  the checkpoint snapshot will be blocked for long-time, in an extreme case, the flink streaming job will never checkpoint successfully if those splits data are huge. 

OK, above is the problem that we want to solve.  

The idea way is:   the mail box executor will only enqueue one split and checkpoint barrier could be inserted at any time.  For example,  we are reading the records from `split1`. At the same time, there're more splits coming from upstream operator :  `split2`, `split3`, `split4`,  then the three splits will be buffered in the in-memory queue ( the variable `Queue<FlinkInputSplit> splits` in PR).   The current flink network queue is empty because we've just buffered those `split2`, `split3`, `split4` in `splits` queue when `processElement`.    Now the checkpoint barrier comes,   the operator read it from network queue and put it in mailbox,  then finally we could trigger the checkpoint snapshot once finished reading  records from `splits`,  don't have to wait all `split2`, `split3`, `split4` 's records reading.

Reply:
After reconsider the code design,   I think the current version is the correct one: https://github.com/apache/iceberg/pull/1793/commits/33554971c192cd6b94f5d7ae7766219cd7e7b6d9

Reply:
> I also don't see where the checkpoint lock is passed to the executor, so how is the executor holding the checkpoint lock?

After the checkpoint lock question,  Flink 1.10.0 has changed the checkpoint-lock model to a mailbox-based model (https://issues.apache.org/jira/browse/FLINK-12477), means we don't have to grab the checkpoint lock when coming a checkpoint barrier because we only have one thread to consume from mailbox so we could just enqueue the checkpoint action in the mailbox and wait for being invoked.  (this https://github.com/apache/iceberg/pull/1793#discussion_r553304000 is actually stale now).    The [SourceStreamTask](https://github.com/apache/flink/blob/7aafc4c51b0314efbfabb94c5efea330914a60c7/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java#L80) still retain the checkpoint lock because there're lots of legacy flink connector which are using this checkpoint lock way to synchronized between normal records and checkpoint barrier.

Reply:
I thought that `MailboxExecutor` ran a separate thread. If so, how is it automatically synchronized with the thread running `processElement`?

Reply:
Let's discuss in this comment .https://github.com/apache/iceberg/pull/1793/files#r556249432